SU2_CFD/src/solvers/CSpeciesFlameletSolver.cpp

```diff
@@ -0,0 +1,823 @@
+
+    su2double dRhodT = 0.0;
+    if (variable_density) {
+      dRhodT = -Density / Temperature;
+    }
+
+    /*--- Passive scalars have no impact on the density. ---*/
+
+    su2double dRhodC = 0.0;
+
+    /*--- Modify matrix diagonal with term including volume and time step. ---*/
+
+    su2double Vol = geometry->nodes->GetVolume(iPoint);
+    su2double Delta = Vol / (config->GetCFLRedCoeff_Species() * solver_container[FLOW_SOL]->GetNodes()->GetDelta_Time(iPoint));
+
+    /*--- Calculating the inverse of the preconditioning matrix
+     that multiplies the time derivative during time integration. ---*/
+
+    if (implicit) {
+      for (unsigned short iVar = 0; iVar < nVar; iVar++) {
+        unsigned long total_index = iPoint * nVar + iVar;
+
+        su2double scalar = nodes->GetSolution(iPoint, iVar);
+
+        /*--- Compute the lag terms for the decoupled linear system from
+         the mean flow equations and add to the residual for the scalar.
+         In short, we are effectively making these terms explicit. ---*/
+
+        su2double artcompc1 = SolP * scalar / (Density * BetaInc2);
+        su2double artcompc2 = SolT * dRhodT * scalar / (Density);
+
+        LinSysRes[total_index] += artcompc1 + artcompc2;
+
+        /*--- Add the extra Jacobian term to the scalar system. ---*/
+
+        su2double Jaccomp = scalar * dRhodC + Density;
+        su2double JacTerm = Jaccomp * Delta;
+
+        Jacobian.AddVal2Diag(iPoint, JacTerm);
+      }
+    }
+  }
+}
+
+void CSpeciesFlameletSolver::Source_Residual(CGeometry* geometry, CSolver** solver_container,
+                                             CNumerics** numerics_container, CConfig* config, unsigned short iMesh) {
+  const bool implicit = (config->GetKind_TimeIntScheme_Flow() == EULER_IMPLICIT);
+  const bool axisymmetric = config->GetAxisymmetric();
+
+  auto* flowNodes = su2staticcast_p<CFlowVariable*>(solver_container[FLOW_SOL]->GetNodes());
+  auto* first_numerics = numerics_container[SOURCE_FIRST_TERM + omp_get_thread_num() * MAX_TERMS];
+
+  SU2_OMP_FOR_DYN(omp_chunk_size)
+  for (auto i_point = 0u; i_point < nPointDomain; i_point++) {
+    /*--- Set primitive variables w/o reconstruction ---*/
+
+    first_numerics->SetPrimitive(flowNodes->GetPrimitive(i_point), nullptr);
+
+    /*--- Set scalar variables w/o reconstruction ---*/
+
+    first_numerics->SetScalarVar(nodes->GetSolution(i_point), nullptr);
+
+    first_numerics->SetDiffusionCoeff(nodes->GetDiffusivity(i_point), nodes->GetDiffusivity(i_point));
+
+    /*--- Set volume of the dual cell. ---*/
+
+    first_numerics->SetVolume(geometry->nodes->GetVolume(i_point));
+
+    /*--- Retrieve scalar sources from CVariable class and update numerics class data. ---*/
+    first_numerics->SetScalarSources(nodes->GetScalarSources(i_point));
+
+    auto residual = first_numerics->ComputeResidual(config);
+
+    /*--- Add Residual ---*/
+
+    LinSysRes.SubtractBlock(i_point, residual);
```

Comment:
This and the source numerics class is quite a bit of boiler plate to essentially just do "residual -= volume * sources", please consider simplifying to add the residual directly (less code, more efficient, and you keep the position in the numerics container free for something more special).

Reply:
Superseded by the comment about using the table in numerics